src/brevitas_examples/llm/llm_quant/learned_round_utils.py

```diff
@@ -0,0 +1,213 @@
+                keep_gpu: bool = True):
+            self.cache_args = cache_args
+            self.cache_kwargs = cache_kwargs
+            self.cache_outs = cache_outs
+
+            self.store_args = store_args
+            self.store_kwargs = store_kwargs
+            self.store_outs = store_outs
+
+            self.keep_gpu = keep_gpu
+
+        def __call__(self, module, args, kwargs, output):
+            # NOTE: If args/kwargs are QuantTensors, should include logic to unpack their values
+            if isinstance(output, (tuple, list)):
+                output = output[0]
+
+            # Store each element in the appropiate cache
+            for element_to_cache, should_cache, cache in zip(
+                [args, kwargs, output],
+                [self.store_args, self.store_kwargs, self.store_outs],
+                [self.cache_args, self.cache_kwargs, self.cache_outs]
+            ):
+                if should_cache:
+                    if not self.keep_gpu:
+                        element_to_cache = send_to_device(element_to_cache, 'cpu')
+                    cache.append(element_to_cache)
+
+            raise StopFwdException
+
+    def _save_inp_out_data(
+            self,
+            model: nn.Module,
+            module: nn.Module,
+            dataloader: DataLoader,
+            cache_args: List,
+            cache_kwargs: List,
+            cache_outs: List,
+            store_args: bool = True,
+            store_kwargs: bool = False,
+            store_outs: bool = True,
+            keep_gpu: bool = True,
+            disable_quant=False) -> None:
+        if disable_quant:
+            disable_quant_class = DisableEnableQuantization()
+            disable_quant_class.disable_act_quantization(model, False)
+            disable_quant_class.disable_param_quantization(model, False)
+            return_quant_tensor_state = disable_return_quant_tensor(model)
+
+        device = next(module.parameters()).device
+        data_saver = LearnedRoundLLMUtils._DataSaverHookLLM(
+            cache_args, cache_kwargs, cache_outs, store_args, store_kwargs, store_outs, keep_gpu)
+        handle = module.register_forward_hook(data_saver, with_kwargs=True)
+        with torch.no_grad():
+            for inps in dataloader:
+                try:
+                    inps = send_to_device(inps, device)
+                    model(**inps)
+                except StopFwdException:
+                    pass
+        handle.remove()
+        if disable_quant:
+            disable_quant_class.enable_act_quantization(model, False)
+            disable_quant_class.enable_param_quantization(model, False)
+            restore_return_quant_tensor(model, return_quant_tensor_state)
+
+    def init_model_learned_round(self, model: nn.Module) -> None:
+        self.llm_cache_state = model.config.use_cache
+        model.config.use_cache = False
+
+    def finish_model_learned_round(self, model: nn.Module) -> None:
```

Comment:
This would be a part of some function implemented in llm utils file, not of the PTQ algorithm

Reply:
This method is not needed in that case